consul/pool.go

```diff
@@ -280,24 +288,51 @@ func (p *ConnPool) DialTimeout(dc string, addr net.Addr, timeout time.Duration)
 
 	// Check if TLS is enabled
 	if p.tlsWrap != nil {
-		// Switch the connection into TLS mode
-		if _, err := conn.Write([]byte{byte(rpcTLS)}); err != nil {
-			conn.Close()
+		doWrap, err := memberExpectsTLS(addr.String(), p.memberFunc())
+		if err != nil {
 			return nil, nil, err
 		}
 
-		// Wrap the connection in a TLS client
-		tlsConn, err := p.tlsWrap(dc, conn)
-		if err != nil {
-			conn.Close()
-			return nil, nil, err
+		if doWrap {
+			// Switch the connection into TLS mode
+			if _, err := conn.Write([]byte{byte(rpcTLS)}); err != nil {
+				conn.Close()
+				return nil, nil, err
+			}
+
+			// Wrap the connection in a TLS client
+			tlsConn, err := p.tlsWrap(dc, conn)
+			if err != nil {
+				conn.Close()
+				return nil, nil, err
+			}
+			conn = tlsConn
 		}
-		conn = tlsConn
 	}
 
 	return conn, hc, nil
 }
 
+// memberExpectsTLS returns true if the member with the specified address is
+// verifying incoming TLS connections
+func memberExpectsTLS(addr string, members []serf.Member) (bool, error) {
```

Comment:
This looks pretty expensive since it's scanning all the members. In other methods where we use the `connPool`, we actually have an `agent.Server`, which is essentially the Serf info - see https://github.com/hashicorp/consul/blob/master/consul/rpc.go#L259-L266. We could probably put the TLS logic in there, and pass that through.